File: algorithms/searching/binary-search/binary_search.go
package search

func BinarySearch(item int, arr []int) int {
	length := len(arr)
	low := 0
	high := length - 1

	for low <= high {
		mid := low + (high-low)/2
		if item == arr[mid] {
			return item
		} else if item < arr[mid] {
			high = mid - 1
		} else if item > arr[mid] {
			low = mid + 1
		}
	}
	return -1
}

File: computer-sience/algorithms/searching/binary-search/cpp_tests/binary_search.cpp
#include "binary_search.h"
#include <catch2/catch_test_macros.hpp>

TEST_CASE("binary_search", "[lower_bound]") {
    std::vector<int> arr {1, 2, 3, 4, 5, 6, 6, 7, 8, 9, 10}; 
    REQUIRE(Search::lower_bound(6, arr) == 5);
}

